### PR TITLE
Fixes appengine-web.xml generation in Ultimate when the web resource dir does not exist

### PR DESCRIPTION
**Pull request by etanshaul**
Progress towards #1948 and #2191 

Updates the `appengine-web.xml` parent directory discovery logic for Ultimate Edition as follows:
- Added: If the web resource directory has a path specified, but that path does not map to a valid directory (which is the case when you import a vanilla Maven or Gradle Spring boot project), it will create that directory recursively (including all of the parent directories, under the project)
- Fixed: If there are no WEB-INF folders found within the available web-roots, instead of just choosing the first one to try and create a WEB-INF, now it will first try and find one that is a valid directory

In addition to new tests, I also cleaned up existing ones as they were pretty hard to follow before.

Example of what happens when importing Spring Boot Maven project now:

1) create a plain Maven WAR Spring Boot project (I used the initializr):
screen shot 2018-06-27 at 9 10 43 pm

2) project is created. it is not yet an app-engine project:
screen shot 2018-06-27 at 9 11 40 pm

3) Notice the defaulted web resource directory on the web facet. This directory does not yet exist:
screen shot 2018-06-27 at 9 12 43 pm

4) Add appengine-standard support via the cloud tools menu. Notice that the appengine-web.xml is created in the correct location (previously it was not created at all):
screen shot 2018-06-27 at 9 14 01 pm

Now deploy and local run work right away, avoiding the error in #2191

### CODE REVIEW

**Comment by etanshaul** [comment]
@patflynn fyi

**Comment by etanshaul** [comment]
note there are still a couple things left to do to fully close the two mentions issues:

- have a appengine-web.xml generation strategy for Community (where we cannot rely on the web facet)
- have a robust fix for #2191 that involves proper error messaging to the user (this PR doesn't deal with handling the error, just goes a step towards avoid it in the first place)

**Comment by patflynn** [comment]
@etanshaul wow, that was fast. :+1: 

@saturnism FYI

File: app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java [hunk 1]
```diff
@@ -96,26 +107,38 @@ public VirtualFile suggestParentDirectoryForAppEngineWebXml(
...
+          return webInfDir;
         }
       }
     }
 
     try {
-      VirtualFile webRootFile = webRoots.get(0).getFile();
-      if (webRootFile != null) {
-        return VfsUtil.createDirectoryIfMissing(webRootFile, WEB_INF);
+      Optional<VirtualFile> webRootDir =
```
Comment (ivanporty) [inline app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java]: so here we're looking for a web root file that exists (physically) on the disk but doesn't contain WEB-INF, after initial loop on `webRoots`, correct?

File: app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java [hunk 2]
```diff
@@ -96,26 +107,38 @@ public VirtualFile suggestParentDirectoryForAppEngineWebXml(
...
+          webRoots.stream().map(WebRoot::getFile).filter(Objects::nonNull).findFirst();
+
+      if (webRootDir.isPresent()) {
+        return VfsUtil.createDirectoryIfMissing(webRootDir.get(), WEB_INF);
+      } else {
+        // There is at least one web-root path, but none that exist, select the first and create it
+        String presentableUrl = webRoots.get(0).getPresentableUrl();
```
Comment (ivanporty) [inline app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java]: What does this return exactly (`getPresentableUrl()`), absolute path in the file system?

Comment (etanshaul) [reply app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java]: yes it does

File: app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java (hunk 1 above)
Comment (etanshaul) [reply app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java]: exactly. previously it was just grabbing the first one which may or may not have physically existed, which was not optimal

Comment (ivanporty) [reply app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java]: could we add a small comment here to define a bit better what it means when a VirtualFile exists - meaning the root exists and we don't need to create it?

File: app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java [hunk 3]
```diff
@@ -96,26 +107,38 @@ public VirtualFile suggestParentDirectoryForAppEngineWebXml(
...
+        if (presentableUrl != null) {
+          VirtualFile newWebRoot = VfsUtil.createDirectories(presentableUrl);
+
+          if (newWebRoot != null) {
+            return VfsUtil.createDirectoryIfMissing(newWebRoot, WEB_INF);
```
Comment (ivanporty) [inline app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java]: identical line with 128, could we use the same path possibly?

Comment (etanshaul) [reply app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java]: done
